src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs

```diff
@@ -131,13 +201,128 @@ private static bool ApplyChangesFromTC25ThroughTC28(JObject sarifLog)
                     if (columnKind == null)
                     {
                         run["columnKind"] = "utf16CodeUnits";
-                    }                
+                    }
                 }
             }
 
             return modifiedLog;
         }
 
+        private static JArray ConstructLogicalLocationsArray(
+            JObject logicalLocations,
+            Dictionary<JObject, int> jObjectToIndexMap,
+            Dictionary<LogicalLocation, int> logicalLocationToIndexMap,
+            out Dictionary<string, int> fullyQualifiedLogicalNameToIndexMap)
+        {
+            fullyQualifiedLogicalNameToIndexMap = null;
+
+            if (logicalLocations == null) { return null; }
+
+            logicalLocationToIndexMap = new Dictionary<LogicalLocation, int>(LogicalLocation.ValueComparer);
+
+            fullyQualifiedLogicalNameToIndexMap = new Dictionary<string, int>();
+
+            foreach (JProperty logicalLocationEntry in logicalLocations.Properties())
+            {
+                // This condition may occur if we've already procssed a parent logical
+                // location that happened to be enumerated after one of its children
+                if (fullyQualifiedLogicalNameToIndexMap.ContainsKey(logicalLocationEntry.Name)) { continue; }
+
+                fullyQualifiedLogicalNameToIndexMap = CreateLogicalLocationArrayEntry(
+                    logicalLocations,
+                    logicalLocationEntry.Name, 
+                    (JObject)logicalLocationEntry.Value,
+                    logicalLocationToIndexMap, 
+                    jObjectToIndexMap,
+                    fullyQualifiedLogicalNameToIndexMap);
+            }
+
+            var updatedArrayElements = new JObject[jObjectToIndexMap.Count];
+
+            foreach (KeyValuePair<JObject, int> keyValuePair in jObjectToIndexMap)
+            {
+                int index = keyValuePair.Value;
+                JObject updatedLogicalLocation = keyValuePair.Key;
+                updatedArrayElements[index] = updatedLogicalLocation;
+            }
+
+            return new JArray(updatedArrayElements);
+        }
+
+        private static Dictionary<string, int> CreateLogicalLocationArrayEntry(
+            JObject logicalLocationsDictionary,
+            string keyName,
+            JObject logicalLocation,
+            Dictionary<LogicalLocation, int> logicalLocationToIndexMap,
+            Dictionary<JObject, int> jObjectToIndexMap,
+            Dictionary<string, int> keyToIndexMap)
+        {
+            int index;
+            keyToIndexMap = keyToIndexMap ?? new Dictionary<string, int>();
```

Comment:
>keyToIndexMap = keyToIndexMap ?? new Dictionary<string, int>(); [](start = 12, length = 63)

The caller is passing a non-null dictionary. IMO you should just update it in place and not bother to return a "new" one (which is only necessary if the input had been `null`). I'd replace this with an assertion that `keyToIndexMap` is non-null, update it in place, and return `void`.

And the method name is a bit confusing: a method `CreateThing` should return a `Thing`. This method doesn't do that, whether or not you adopt my suggestion around `keyToIndexMap`. It should be `AddEntryToFullyQualifiedNameToIndexMap`, and the signature IMO should be (putting "the thing we're adding" first in the argument list, and the thing we're modifying last):

```
private static void AddEntryToFullyQualifiedNameToIndexMap(
    string fullyQualifiedLogicalName,
    JObject logicalLocationObject,
    Dictionary<LogicalLocation, int> logicalLocationToIndexMap,
    Dictionary<JObject, int> jObjectToIndexMap,
    Dictionary<string, int> fullyQualifiedLogicalNameToIndexMap);
```

Reply:
These are both good suggestions. 

---
In reply to: [243857089](https://github.com/Microsoft/sarif-sdk/pull/1170#discussion_r243857089) [](ancestors = 243857089)

Reply:
Fixed by Michael in PR #1184.

---
In reply to: [244051988](https://github.com/Microsoft/sarif-sdk/pull/1170#discussion_r244051988) [](ancestors = 244051988,243857089)